File: app/src/main/java/ilapin/opengl_research/domain/ScrollController.kt
package ilapin.opengl_research.domain

import ilapin.common.input.TouchEvent
import ilapin.common.kotlin.safeLet

/**
 * @author raynor on 07.02.20.
 */
class ScrollController(private val touchEventsRepository: TouchEventsRepository) {

    private var prevTouchEvent: TouchEvent? = null

    var scrollEvent: ScrollEvent? = null

    fun update() {
        scrollEvent = null
        val touchEvent = touchEventsRepository.touchEvents.firstOrNull()
        safeLet(prevTouchEvent, touchEvent) { prev, current ->
            scrollEvent = ScrollEvent(current.x - prev.x, current.y - prev.y)
        }
        prevTouchEvent = touchEvent
    }
}

File: app/src/main/assets/scenes/scroll_controller.js
function ScrollController(gestureConsumer) {

    this.gestureConsumer = gestureConsumer;

    this.prevTouchEvent = null;

    this.scrollEvent = null;

    this.update = function() {
        this.scrollEvent = null;

        var touchEvents = this.gestureConsumer.touchEvents;

        if (touchEvents.size() == 0) {
            return;
        }

        for (var iterator = touchEvents.iterator(); iterator.hasNext();) {
            var touchEvent = iterator.next();

            if (this.prevTouchEvent != null) {
                this.scrollEvent = new ScrollEvent(
                    touchEvent.x - this.prevTouchEvent.x,
                    touchEvent.y - this.prevTouchEvent.y
                );
            }

            if (
                touchEvent.action == Packages.ilapin.common.input.TouchEvent.Action.UP ||
                touchEvent.action == Packages.ilapin.common.input.TouchEvent.Action.CANCEL
            ) {
                this.prevTouchEvent = null;
            } else {
                this.prevTouchEvent = touchEvent;
            }
        }
    };
}

function ScrollEvent(dx, dy) {
    this.dx = dx;
    this.dy = dy;
}